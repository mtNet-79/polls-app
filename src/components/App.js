import { handleInitialData } from "../actions/shared";
import PollPage from "./PollPage";
import { useState, useEffect, Fragment } from "react";
import Dashboard from "./Dashboard";
import { connect } from "react-redux";
import Nav from './Nav';
import LogIn from "./LogIn";
import Leaderboard from "./Leaderboard";
import AddUser from "./AddUser";
import LoadingBar from "react-redux-loading-bar";
import { Routes, Route } from "react-router-dom";
import CreatePoll from "./CreatePoll";
import { useLocation, useNavigate, useParams } from "react-router-dom";

const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};

const App = (props) => {
  console.log("app props: ", props);
  const [image, setImage] = useState("");
  useEffect(() => {
   
    const generateRandomAvatar = () => {
      fetch("https://avatar-endpoint.herokuapp.com/api/").then((res) => {
        res.blob().then((blobRes) => {
          const data = blobRes;
          const urlCreator = window.URL || window.webkitURL;
          const urlObj = urlCreator.createObjectURL(data);
          setImage(urlObj);
          console.log("trigged in generateavatar");
          props.dispatch(handleInitialData());
        });
      });
    }

    const alreadyLoggedIn = () => {
      console.log("triggerd at already")
      props.dispatch(handleInitialData());
    }
    
    props.authedUser === null 
      ? generateRandomAvatar() 
      : alreadyLoggedIn();
  }, []);



  return (
    <Fragment>
      <LoadingBar />       
        {props.loading ? null : (   
            <div className="container">              
              {props.location.pathname === "/add-user" 
                ? <AddUser image={image} />
                : props.authedUser
                  ? (
                    <Fragment>                   
                      <Nav />
                      <Routes>
                        <Route path="/" exact element={<Dashboard />} />                        
                        <Route path="/login" element={<LogIn image={image} />} />
                        <Route path="/new" element={<CreatePoll />} />
                        <Route path="/leaderboard" element={<Leaderboard />} />
                        <Route path="/questions/:pid" element={<PollPage image={image}/>} />
                      </Routes>
                    </Fragment>   
                  )           
                  : <LogIn image={image} />                  
              }            
            </div>
        )}
    </Fragment>
  );
};

const mapStateToProps = ({  authedUser, loading, users }, props) => {  
  // const keys = Object.keys(users);
  const { location } = props.router;
  return {
    loading: loading,
    authedUser,
    avatarURL: users[authedUser]?.avatarURL,
    location,
  }
    
    
};
  

export default withRouter(connect(mapStateToProps)(App));
