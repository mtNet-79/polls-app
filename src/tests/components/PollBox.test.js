import { screen, waitFor, within } from "@testing-library/react";
// import PollContainer from "../../Components/PollContainer";
import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import App from "../../Components/App";
import { renderWithProviders } from "./test-utils";

//unit tests
test("all elements are rendered showing unaswered polls / take snapshot", async () => {
  const {store, container} =renderWithProviders(<App />, {route: '/', authedUser: 'mthornton'});

  await waitFor(
    () => expect(screen.getByText(/new polls/i)).toBeInTheDocument(),
    {
        timeout: 2000,
    }
);


 //all unanswered polls are present 
 const dashboardList = screen.getAllByRole('list')[2]
const pollBoxes = within(dashboardList).getAllByRole("listitem");


const answeredPolls = Object.keys(store.getState().users[store.getState().authedUser].answers);  
const pollsCount = Object.keys(store.getState().polls).length;  
expect(pollBoxes.length).toBe(pollsCount - answeredPolls.length)

const pollBox = pollBoxes[0]
//  expect(within(pollBox).getByText('mtsamis')).toBeInTheDocument()
//  expect(within(pollBox).getByText(/1:16:pm \| 4\/30\/2017/i)).toBeInTheDocument()
 expect(within(pollBox).getByRole('button', {name: /show poll/i})).toBeInTheDocument()

 expect(container).toMatchSnapshot();
 
});
