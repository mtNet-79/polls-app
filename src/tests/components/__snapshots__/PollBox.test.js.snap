// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all elements are rendered showing unaswered polls / take snapshot 1`] = `
<div>
  <div
    style="direction: ltr;"
  >
    <div
      class=""
      style="width: 100%; transition: width 200ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 1;"
    />
    <div
      style="display: table; clear: both;"
    />
  </div>
  <div
    class="container"
  >
    <div
      class="nav"
    >
      <ul>
        <li>
          <a
            class=""
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class=""
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li>
          <a
            class=""
            href="/new"
          >
            Post
          </a>
        </li>
      </ul>
      <ul>
        <li>
          <img
            alt="icon"
            class="avatar-small"
            src="https://avatar-endpoint.herokuapp.com/api/"
          />
          <span>
            mthornton
          </span>
        </li>
        <li>
          <a
            href="/login"
          >
            Log out
          </a>
        </li>
      </ul>
    </div>
    <div
      class="dashboard"
    >
      <div
        class="poll-container"
      >
        <div
          class="flex-row"
        >
          <div
            class="cell"
          >
            <input
              checked=""
              class="react-switch-checkbox"
              id="react-switch-new"
              type="checkbox"
            />
            <label
              class="react-switch-label"
              for="react-switch-new"
            >
              <span
                class="react-switch-button"
              />
            </label>
          </div>
          <div
            class="cell"
          >
            <h2
              class="polls-header center"
            >
              New Polls
            </h2>
          </div>
        </div>
        <ul
          class="flex-row-wrap"
        >
          <li>
            <div
              class="poll-box"
            >
              <div>
                mtsamis
              </div>
              <div>
                1:16:PM | 4/30/2017
              </div>
              <button
                class="btn"
              >
                Show Poll
              </button>
            </div>
          </li>
          <li>
            <div
              class="poll-box"
            >
              <div>
                tylermcginnis
              </div>
              <div>
                6:09:AM | 3/15/2017
              </div>
              <button
                class="btn"
              >
                Show Poll
              </button>
            </div>
          </li>
          <li>
            <div
              class="poll-box"
            >
              <div>
                sarahedo
              </div>
              <div>
                3:22:PM | 3/3/2017
              </div>
              <button
                class="btn"
              >
                Show Poll
              </button>
            </div>
          </li>
          <li>
            <div
              class="poll-box"
            >
              <div>
                tylermcginnis
              </div>
              <div>
                4:42:AM | 12/24/2016
              </div>
              <button
                class="btn"
              >
                Show Poll
              </button>
            </div>
          </li>
          <li>
            <div
              class="poll-box"
            >
              <div>
                mtsamis
              </div>
              <div>
                1:02:AM | 7/14/2016
              </div>
              <button
                class="btn"
              >
                Show Poll
              </button>
            </div>
          </li>
          <li>
            <div
              class="poll-box"
            >
              <div>
                sarahedo
              </div>
              <div>
                8:21:PM | 6/28/2016
              </div>
              <button
                class="btn"
              >
                Show Poll
              </button>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
